internal/mongodoc/resource.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package mongodoc
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	"gopkg.in/errgo.v1"
+	"gopkg.in/juju/charm.v6-unstable"
+	"gopkg.in/juju/charm.v6-unstable/resource"
+	"gopkg.in/juju/charmrepo.v2-unstable/csclient/params"
+	"gopkg.in/mgo.v2/bson"
+)
+
+// NewResourceQuery formats the provided details into a mongo query for
+// the identified resource.
+func NewResourceQuery(cURL *charm.URL, resName string, revision int) bson.D {
+	copied := *cURL
+	cURL = &copied
+	cURL.Revision = -1
+	cURL.Series = ""
+	query := bson.D{
+		{"unresolved-charm-url", cURL},
+		{"name", resName},
+	}
+	if revision >= 0 {
+		query = append(query, bson.DocElem{"revision", revision})
+	}
+	return query
+}
+
+// Resource holds the in-database representation of a charm resource
+// at a particular revision.
```

Comment:
Perhaps mention that the (url, name, revision) triple forms the primary key tuple?